### ISSUE TITLE
[bug] ERROR: There are 2 binaries with name buildreq

### ISSUE THREAD

**Issue by spoofedex** [conan-io/conan#7141]
There seems to be an issue using lockfiles, as described in "[How to use lockfiles in CI](https://docs.conan.io/en/latest/versioning/lockfiles.html#how-to-use-lockfiles-in-ci)", in combination with build requirements: when multiple dependencies have the same build requirement, it is listed multiple times in the lockfile, which causes an error when building that package with the generated lockfile.

A question has been raised regarding this (see #6065), but I felt a new bug report was in order: it was in the form of a question, it lacks an easy reproduction, and it is mentioned that it is caused by having a dependency as both a build requirement as a requirement, which is not (explicitly) the case here.

I'm not aware of an actual workaround. If the package is not built, then the error will not occur (even though the dependency is still listed twice). As the build order is determined using "conan graph build-order", the build will be skipped if it has already been created. Hence, by building the duplicate packages manually, it does succeed (for now).

Note that as far as I am aware, I am following the documentation exactly here. Maybe I'm not supposed to add the build parameters to "graph lock", but then I can't even determine the build-order because buildreq isn't in the lockfile. So I feel this should be a common and regular use case for Conan.

### Environment Details (include every applicable attribute)
  * Operating System+version: Windows 10
  * Compiler+version: not applicable (nothing being compiled)
  * Conan version: reproduced with 1.25.2 and 1.24
  * Python version: reproduced with 3.7.3 and 3.8.3

### Steps to reproduce (Include if Applicable)
Given three Conan packages, with the following conanfile.py:
buildreq/conanfile.py:
```python
import os
from conans import ConanFile

class BuildReqConan(ConanFile):
    name = "buildreq"
    version = "1.0.0"
```

PkgA/conanfile.py:
```python
from conans import ConanFile

class PkgAConan(ConanFile):
    name = "PkgA"
    version = "1.0.0"
    
    def build_requirements(self):
        self.build_requires("buildreq/1.0.0@test/test")
```

PkgB/conanfile.py:
```python
from conans import ConanFile

class PkgBConan(ConanFile):
    name = "PkgB"
    version = "1.0.0"
    requires = "PkgA/1.0.0@test/test"
    
    def build_requirements(self):
        self.build_requires("buildreq/1.0.0@test/test")
```

1. Export the three packages (`conan export . test/test`)
2. Create a lockfile: `conan graph lock --lockfile . PkgB/1.0.0@test/test --build missing --build outdated --build cascade`
3. Note: in the created lockfile, buildreq is listed twice. This seems to be an error state already.
4. Build "buildreq" with the lockfile: `conan install --lockfile . buildreq/1.0.0@test/test`
5. Result: "ERROR: There are 2 binaries with name buildreq"

Note that these steps are similar to that described in the official documentation in section "[How to use lockfiles in CI](https://docs.conan.io/en/latest/versioning/lockfiles.html#how-to-use-lockfiles-in-ci)".

Best regards,
SpoofedEx

**Comment by spoofedex** [comment]
The duplicate dependencies are added in GraphManager::_recurse_build_requires. It seems that normally DepsGraphBuilder::_expand_require will filter out duplicates, but it only considers the public dependencies and public closure, which (apparently) does not contain build requirements.

**Comment by jgsogo** [comment]
I've opened a BR with a minimal test reproducing the error. Check #7144

**Comment by memsharded** [comment]
This has been addressed by #7243, that does big changes for the lockfiles. They will be released in Conan 1.28, please keep tuned to the release, and let us know if everything works.
